dxr/query.py

```diff
@@ -108,50 +136,30 @@ def results(self, offset=0, limit=100):
         is_line_query = any(f.domain == LINE for f in
                             chain.from_iterable(filters))
 
-        # An ORed-together ball for each term's filters, omitting filters that
-        # punt by returning {} and ors that contain nothing but punts:
-        ors = filter(None, [filter(None, (f.filter() for f in term))
-                            for term in filters])
-        ors = [{'or': x} for x in ors]
-
-        if not is_line_query:
-            # Don't show folders yet in search results. I don't think the JS
-            # is able to handle them.
-            ors.append({'term': {'is_folder': False}})
-
-        if ors:
-            query = {
-                'filtered': {
-                    'query': {
-                        'match_all': {}
-                    },
-                    'filter': {
-                        'and': ors
-                    }
-                }
-            }
-        else:
-            query = {
-                'match_all': {}
-            }
+        return self._perform_search(filters, is_line_query, offset, limit)
 
-        results = self.es_search(
-            {'query': query,
-             'sort': ['path', 'number'] if is_line_query else ['path'],
-             'from': offset,
-             'size': limit},
-            doc_type=LINE if is_line_query else FILE)['hits']
-        result_count = results['total']
-        results = [r['_source'] for r in results['hits']]
+    # Test: If var-ref (or any structural query) returns 2 refs on one line, they should both get highlit.
 
-        path_highlighters = [f.highlight_path for f in chain.from_iterable(filters)
-                             if hasattr(f, 'highlight_path')]
-        return {'result_count': result_count,
-                'results': self._line_query_results(filters, results, path_highlighters)
-                           if is_line_query
-                           else self._file_query_results(results, path_highlighters)}
+    def promoted_paths(self, term, promote_limit=5):
+        """Return a tuple (total number of path results, promoted paths),
+        where the number of promoted paths is no more than promote_limit."""
 
-        # Test: If var-ref (or any structural query) returns 2 refs on one line, they should both get highlit.
+        # We cannot import PathFilter at top of file because of circular dependency.
+        from dxr.plugins.core import PathFilter
+
+        filters = [[PathFilter(term, self.enabled_plugins)]]
+        path_query = {
+            'constant_score': {'query': _construct_query(filters), 'boost': 0.5}}
+        # We add a second query that boosts exact path part matches.
+        exact_query = {
+            'constant_score': {'query': {'match': {'path.parts': term['arg']}}, 'boost': 2.0}}
```

Comment:
Since you're using a constant_score query here and thus not taking advantage of scoring, is it possible to use an ES filter (rather than an ES query)? Filters are an order of magnitude faster and cacheable.